app/controllers/csv_preview_controller.rb

```diff
@@ -1,13 +1,57 @@
+require "csv"
+
 class CsvPreviewController < ApplicationController
   def show
-    legacy_url_path = request.path.sub("/preview", "").sub(/^\//, "")
     @asset = GdsApi.asset_manager.whitehall_asset(legacy_url_path).to_hash
 
+    maximum_columns = 50
+    maximum_rows = 1000
+    csv_preview = CSV.parse(media, encoding:, headers: true)
+
+    @csv_rows = csv_preview.to_a.map { |row|
+      row.map { |column|
+        {
+          text: column,
+        }
+      }.take(maximum_columns)
+    }.take(maximum_rows + 1)
+
     parent_document_path = URI(@asset["parent_document_url"]).request_uri
     @content_item = GdsApi.content_store.content_item(parent_document_path).to_hash
 
     @attachment_metadata = @content_item.dig("details", "attachments").select do |attachment|
       attachment["url"] =~ /#{legacy_url_path}$/
     end
   end
+
+private
+
+  def legacy_url_path
+    request.path.sub("/preview", "").sub(/^\//, "")
```

Comment:
How stable is this pattern? Where do the path chunks to be substituted come from?

Reply:
The route will [always have `/preview` at the end and a leading slash](https://github.com/alphagov/frontend/blob/3f08e1b5b249aab9d0fbd0a282b6675df0f14df3/config/routes.rb#L109), so I think this is safe to substitute.